gluten-core/src/main/scala/org/apache/spark/shuffle/ShuffleManagerRegistry.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.shuffle
+
+import org.apache.spark.SparkConf
+import org.apache.spark.util.Utils
+
+import scala.collection.mutable
+
+class ShuffleManagerRegistry private[ShuffleManagerRegistry] {
+  import ShuffleManagerRegistry._
+  private val all: mutable.Buffer[(LookupKey, String)] = mutable.Buffer()
+  private val routerBuilders: mutable.Buffer[RouterBuilder] = mutable.Buffer()
```

Comment:
Could you explain in what cases there would be multiple RouterBuilders? Thanks!

Reply:
I think we usually have one single router builder at the same time. So this practice can be simplified in future with some kind of assertions (something like `if (router created) throw error`). Though it's benign at the moment.

Reply:
The reason we don't simply adopt a lazy initialization here is because the instantiation of `RouterBuilder` is parameterized: It takes `SparkConf` and `isDriver` in ctor. So we'd guard our code from the twice invocation with different parameters, no matter it's practically possible or not.